xds/src/main/java/io/grpc/xds/RingHashLoadBalancer.java

```diff
@@ -142,6 +146,19 @@ public void onSubchannelState(ConnectivityStateInfo newState) {
     for (EquivalentAddressGroup addr : removedAddrs) {
       removedSubchannels.add(subchannels.remove(addr));
     }
+    // If we need to proactively start connecting, iterate through all the subchannels, starting
+    // at a random position.
+    // Alternatively, we should better start at the same position.
+    connectionAttemptIterator = subchannels.values().iterator();
+    int randomAdvance = random.nextInt(subchannels.size());
+    while (randomAdvance-- > 0) {
+      if (!connectionAttemptIterator.hasNext()) {
```

Comment:
Why check `hasNext() == false`? This is a brand-new iterator, created on line 152. We know it isn't empty because of line 89. Seems we can just call `next()` repeatedly. We don't even need an assertion check, as `next()` will throw if we try to loop too many times (because of a bug).

Reply:
did you see iterator implementations that `hasNext()` always needs to be called before you call `.next()` to allow it move their pointer? So was just playing safe. 
If we assume it is always protected by syncContext then no it does not need the check. And surfacing the bug may not be bad behavior.

Reply:
You're talking about broken Iterator implementations? Do we suspect the JDK has such a broken implementation? `next()` is well defined without calling `hasNext()`; I think it is fair to depend on the API being properly implemented.